### PR TITLE
Nb/tabular/increase speed outlier detection + timeout

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Major increase in running speed due to vectorization and other cool numpy tricks + Added a timeout mechanism.

It is still not that fast (7 sec for 10K samples with dimension 13).

### CODE REVIEW

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 1]
```diff
@@ -42,71 +49,84 @@ class OutlierSampleDetection(SingleDatasetCheck):
...
     extent_parameter: int, default: 3
         Extend parameter for LoOP algorithm.
-    n_samples : int , default: 50_000
+    n_samples : int , default: 5_000
         number of samples to use for this check.
     n_to_show : int , default: 5
         number of data elements with the highest outlier score to show (out of sample).
     random_state : int, default: 42
         random seed for all check internals.
+    timeout : int, default: 10
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:63]:
```suggestion
    timeout : int, default: 5
```

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 2]
```diff
@@ -42,71 +49,84 @@ class OutlierSampleDetection(SingleDatasetCheck):
...
             columns: Union[Hashable, List[Hashable], None] = None,
             ignore_columns: Union[Hashable, List[Hashable], None] = None,
-            num_nearest_neighbors: int = 10,
+            nearest_neighbors_percent: float = 0.01,
             extent_parameter: int = 3,
-            n_samples: int = 50_000,
+            n_samples: int = 5_000,
             n_to_show: int = 5,
             random_state: int = 42,
+            timeout: int = 10,
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:76]:
```suggestion
            timeout: int = 5,
```

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 3]
```diff
@@ -42,71 +49,84 @@ class OutlierSampleDetection(SingleDatasetCheck):
...
         df = select_from_dataframe(dataset.data, self.columns, self.ignore_columns)
-        if self.num_nearest_neighbors >= len(df):
-            logger.warning('Passed num_nearest_neighbors %s which is greater than the number of samples in the dataset'
-                           , self.num_nearest_neighbors)
-            self.num_nearest_neighbors = len(df) - 1
-
-        # Calculate distances matrix and retrieve nearest neighbors based on distance matrix.
-        df_cols_for_gower = df[dataset.cat_features + dataset.numerical_features]
-        is_categorical_arr = np.array(df_cols_for_gower.columns.map(lambda x: x in dataset.cat_features), dtype=bool)
+        num_neighbours = max(int(self.nearest_neighbors_percent * df.shape[0]), 5)
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:105]: Have this magic number as an upper case constant in the top of the py file
MIN_NUM_NEIGHBOURS

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 4]
```diff
@@ -42,71 +49,84 @@ class OutlierSampleDetection(SingleDatasetCheck):
...
+
+        start_time = time.time()
+        gower_distance.calculate_nearest_neighbours_distances(
+            df[dataset.cat_features].iloc[:100], df[dataset.numerical_features].iloc[:100], 10)
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:109]: Same for these magic numbers

File: deepchecks/utils/gower_distance.py [hunk 5]
```diff
@@ -64,43 +67,78 @@ def gower_matrix_n_closets(data: np.ndarray, cat_features: np.array, num_neighbo
...
+        representing the indexes of the nearest neighbours.
     """
-    if not isinstance(data, np.ndarray):
-        data = np.asarray(data)
+    num_samples = cat_data.shape[0]
+    num_features = cat_data.shape[1] + numeric_data.shape[1]
+    distances, indexes = np.zeros((num_samples, num_neighbours)), np.zeros((num_samples, num_neighbours))
+    # handle categorical - transform to an ordinal numpy array
+    enc = OrdinalEncoder()
+    cat_data = np.nan_to_num(enc.fit_transform(cat_data.fillna(value=np.nan)), nan=-1)
```
Comment (noamzbr) [inline deepchecks/utils/gower_distance.py:89]: Does pandas fillna fill also pd.NaN, pd.NaT, and python None?

File: tests/utils/test_gower_distance.py [hunk 6]
```diff
@@ -88,3 +57,26 @@ def test_mix_columns_full_matrix_with_nulls():
     assert_that(dist[-1], has_length(data.shape[0]))
     assert_that(dist[3], has_item(greater_than(0.01)))
     assert_that(min(dist[0]), less_than_or_equal_to(0))
+
+
+def test_mix_columns_nn_matrix_with_nulls_vectorized():
```
Comment (noamzbr) [inline tests/utils/test_gower_distance.py:62]: Good idea, are gowers dependencies OK?

File: docs/source/checks/tabular/integrity/plot_outlier_sample_detection.py [hunk 7]
```diff
@@ -56,7 +56,7 @@
 # =============
 # We define the num_neighbors and the extent parameters for the LoOP algorithm.
 
-check = OutlierSampleDetection(num_nearest_neighbors=10, extent_parameter=3)
+check = OutlierSampleDetection(nearest_neighbors_percent=10, extent_parameter=3)
```
Comment (noamzbr) [inline docs/source/checks/tabular/integrity/plot_outlier_sample_detection.py:59]: Why we need 10% here?

File: deepchecks/utils/gower_distance.py (hunk 5 above)
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:89]: Yes, can see in tests :)

File: tests/utils/test_gower_distance.py (hunk 6 above)
Comment (Nadav-Barak) [reply tests/utils/test_gower_distance.py:62]: The dependencies were only added to dev-requirements so as far i understood are not downloaded with the package, am I correct? @matanper

File: docs/source/checks/tabular/integrity/plot_outlier_sample_detection.py (hunk 7 above)
Comment (Nadav-Barak) [reply docs/source/checks/tabular/integrity/plot_outlier_sample_detection.py:59]: Forget to change it, good catch

**Comment by noamzbr** [comment]
Re-add the check to the integrity suite.

File: tests/utils/test_gower_distance.py (hunk 6 above)
Comment (noamzbr) [reply tests/utils/test_gower_distance.py:62]: Yep it's right, just making sure that the testing env will be functional.

**Comment by Nadav-Barak** [comment]
Already added

**Pull request merged**